File: backend/student-service/src/report/helpers/discTrait.ts
/**
 * DISC trait resolver - TypeScript mirror of
 * backend/exam-engine/internal/service/disc_trait.go.
 *
 * The exam engine is the single source of truth for `dominant_trait_id`; the
 * report layer should prefer that stored value for the HEADLINE / narrative
 * trait. This helper exists for (a) a fallback when an attempt predates the
 * engine change, and (b) the top-two blend used by the career/compatibility
 * lookups (see below). `resolveDominantFactor` MUST stay byte-for-byte
 * equivalent to the Go resolver - both are validated against the same canonical
 * cases (disc_trait_test.go and validate_disc_trait_ts.js).
 */

const DISC_FACTOR_PRIORITY: Record<string, number> = { C: 0, D: 1, I: 2, S: 3 };

interface FactorScore {
  factor: string;
  score: number;
}

/** Factors present in `scores`, sorted by score desc, tie-break C>D>I>S. */
function sortedFactors(scores: Record<string, number>): FactorScore[] {
  const list: FactorScore[] = [];
  for (const f of ['D', 'I', 'S', 'C']) {
    const v = scores[f];
    if (v === undefined || v === null) continue;
    list.push({ factor: f, score: Number(v) });
  }
  list.sort((a, b) =>
    a.score !== b.score
      ? b.score - a.score
      : DISC_FACTOR_PRIORITY[a.factor] - DISC_FACTOR_PRIORITY[b.factor],
  );
  return list;
}

/**
 * Resolves the HEADLINE trait code from raw DISC factor sums. Returns a single
 * letter ("D"/"I"/"S"/"C") - a Pure Trait - when one DISC dimension is dominant
 * enough, otherwise the top-two blend (e.g. "DI").
 *
 * Rule (total = D+I+S+C): pure when `top*2 >= total` OR `top/2 > every other`;
 * tie-break C>D>I>S; guard when top/total <= 0.
 */
export function resolveDominantFactor(scores: Record<string, number>): string {
  const list = sortedFactors(scores);
  if (list.length === 0) return '';

  const total = list.reduce((sum, x) => sum + x.score, 0);
  const top = list[0];
  if (list.length === 1) return top.factor;

  if (top.score > 0 && total > 0) {
    if (top.score * 2 >= total) return top.factor; // rule (1): >= 50% of total
    if (list.slice(1).every((r) => top.score / 2 > r.score)) return top.factor; // rule (2)
  }

  return top.factor + list[1].factor;
}

/**
 * The top-two DISC blend (two highest factors, tie-break C>D>I>S) - always the
 * 2-letter code, with NO pure-trait override. This is what the existing 12-blend
 * content keyed by `trait_id` expects.
 */
export function topTwoBlend(scores: Record<string, number>): string {
  const list = sortedFactors(scores);
  if (list.length === 0) return '';
  if (list.length === 1) return list[0].factor;
  return list[0].factor + list[1].factor;
}

/**
 * Resolves the HEADLINE trait code from per-student report data, honouring the
 * exam engine as the single source of truth. Order of preference:
 *   1. `dominant_trait_code` - the code the engine already stored (authoritative,
 *      pure-capable). In production this is always present for completed attempts.
 *   2. the raw DISC sums in `most_answered_answer_type` (these COUNT values ARE the
 *      engine's `disc_scores` input - see groupReportHelper), applying the same
 *      rule the engine uses, so the fallback matches the engine byte-for-byte.
 *   3. the percentage-scaled `score_*` as a last resort (e.g. legacy mock data
 *      with no counts) - ordering is preserved even if the absolute cutoff is not.
 */
export function resolveHeadlineFromReportData(data: {
  dominant_trait_code?: string;
  most_answered_answer_type?: { ANSWER_TYPE: string; COUNT: number }[];
  score_D?: number;
  score_I?: number;
  score_S?: number;
  score_C?: number;
}): string {
  if (data.dominant_trait_code) return data.dominant_trait_code;

  const counts = data.most_answered_answer_type;
  if (counts && counts.length >= 4) {
    const m: Record<string, number> = {};
    for (const c of counts) m[c.ANSWER_TYPE] = Number(c.COUNT);
    return resolveDominantFactor(m);
  }

  return resolveDominantFactor({
    D: data.score_D ?? 0,
    I: data.score_I ?? 0,
    S: data.score_S ?? 0,
    C: data.score_C ?? 0,
  });
}

/** True when a resolved code is a single-letter Pure Trait (D/I/S/C). */
export function isPureTrait(code: string): boolean {
  return code.length === 1 && code !== '';
}

/**
 * Splits a resolved code into [primary, secondary]. For a Pure Trait the
 * secondary is '' - callers that look up two-letter maps should guard on that.
 */
export function splitTrait(code: string): [string, string] {
  return [code.charAt(0) || '', code.charAt(1) || ''];
}

/**
 * Whether the pure traits have their own rows in the `trait_id`-keyed content
 * tables (`career_roles`, `career_role_*`, `trait_based_course_details`).
 *
 * While `false`, a Pure-Trait student's career-guidance and course-compatibility
 * sections fall back to their **top-two blend** (their two highest factors) - the
 * blend rows always exist. Flip to `true` once pure rows are authored to switch
 * pure students onto their own data; nothing else changes.
 */
export const PURE_CAREER_ROWS_AVAILABLE = false;

/**
 * The trait code to use for the `trait_id`-keyed career/compatibility lookups,
 * given the resolved headline trait. This is the single seam where "pure traits
 * borrow the top-two blend's data" lives - see PURE_CAREER_ROWS_AVAILABLE.
 */
export function careerLookupCode(
  resolvedTrait: string,
  scores: Record<string, number>,
): string {
  if (PURE_CAREER_ROWS_AVAILABLE && isPureTrait(resolvedTrait)) {
    return resolvedTrait;
  }
  return topTwoBlend(scores);
}

File: backend/exam-engine/internal/service/disc_trait.go
package service

import "sort"

// discFactorPriority is the deterministic tie-break order C > D > I > S, applied
// when two factors share the same score. It mirrors getTopTwoTraits in the
// report layer and calculateDiscProfile in the specialization report so all
// three trait-resolution paths agree.
var discFactorPriority = map[string]int{"C": 0, "D": 1, "I": 2, "S": 3}

// ResolveDominantFactor turns raw Level-1 DISC factor sums into the dominant
// trait code that is stored on assessment_attempts.dominant_trait_id (via the
// personality_traits.code lookup).
//
// It returns a single letter ("D"/"I"/"S"/"C") - a "Pure Trait" - when one
// dimension is dominant enough, otherwise the top-two blend (e.g. "DI").
//
// Pure-trait rule (raw disc_scores; total = D+I+S+C):
//
//	(1) top*2 >= total          // top is at least 50% of the candidate's own total
//	(2) top/2  >  every other   // top strictly dominates each other dimension
//
// Guard: when the top score or the total is non-positive we never override -
// the result is the same blend (or single factor) the engine produced before,
// so an empty/abandoned attempt is unaffected.
//
// `scores` is the factor->sum map (e.g. {"D":22,"I":16,"S":6,"C":8}); any extra
// keys (such as "total") are ignored. Missing factors are treated as absent
// (i.e. score 0 and excluded from the total), matching the SQL aggregation that
// only returns factors the candidate actually scored on.
func ResolveDominantFactor(scores map[string]float64) string {
	type factorScore struct {
		factor string
		score  float64
	}

	list := make([]factorScore, 0, 4)
	var total float64
	for _, f := range []string{"D", "I", "S", "C"} {
		v, ok := scores[f]
		if !ok {
			continue
		}
		list = append(list, factorScore{factor: f, score: v})
		total += v
	}

	if len(list) == 0 {
		return ""
	}

	sort.SliceStable(list, func(i, j int) bool {
		if list[i].score != list[j].score {
			return list[i].score > list[j].score
		}
		return discFactorPriority[list[i].factor] < discFactorPriority[list[j].factor]
	})

	top := list[0]
	if len(list) == 1 {
		return top.factor
	}

	if top.score > 0 && total > 0 {
		// Rule (1): top is at least half of the whole distribution.
		if top.score*2 >= total {
			return top.factor
		}
		// Rule (2): top more than doubles every other dimension individually.
		dominatesAll := true
		for _, r := range list[1:] {
			if !(top.score/2 > r.score) {
				dominatesAll = false
				break
			}
		}
		if dominatesAll {
			return top.factor
		}
	}

	// Standard dual-trait blend (unchanged 12-combination behaviour).
	return top.factor + list[1].factor
}
